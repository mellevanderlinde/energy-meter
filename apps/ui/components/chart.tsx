"use client";

import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import {
  type ChartConfig,
  ChartContainer,
  ChartTooltip,
  ChartTooltipContent,
} from "@/components/ui/chart";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";
import { type GetDataOutput, getData } from "get-data";
import { type ReactNode, useEffect, useMemo, useState } from "react";
import { Bar, BarChart, CartesianGrid, XAxis, YAxis } from "recharts";

const chartConfig = {
  kwh: {
    label: "kWh",
    color: "hsl(var(--chart-1))",
  },
} satisfies ChartConfig;

export function Chart(): ReactNode {
  const [numberOfDays, setNumberOfDays] = useState("14");
  const [data, setData] = useState<GetDataOutput>([]);

  useEffect(() => {
    async function loadData() {
      const result = await getData(Number.parseFloat(numberOfDays));
      setData(result);
    }
    loadData();
  }, [numberOfDays]);

  const sumPerDay = useMemo(() => {
    const result: { [key: string]: number } = {};
    for (const entry of data) {
      const date = new Date(entry.datetime).toLocaleDateString("en-US");
      if (!result[date]) {
        result[date] = 0;
      }
      result[date] += entry.kwh;
    }
    return Object.entries(result).map(([date, kwh]) => ({ date, kwh }));
  }, [data]);

  return (
    <Card>
      <CardHeader className="flex items-center gap-2 space-y-0 border-b py-5 sm:flex-row">
        <div className="grid flex-1 gap-1 text-center sm:text-left">
          <CardTitle>Energy Usage</CardTitle>
          <CardDescription>Showing kWh consumption over time</CardDescription>
        </div>
        <Select value={numberOfDays} onValueChange={setNumberOfDays}>
          <SelectTrigger
            className="w-[160px] rounded-lg sm:ml-auto"
            aria-label="Select a value"
          >
            <SelectValue placeholder="Last 3 months" />
          </SelectTrigger>
          <SelectContent className="rounded-xl">
            {[
              { value: "14", label: "Last 14 days" },
              { value: "30", label: "Last 30 days" },
              { value: "90", label: "Last 3 months" },
            ].map(({ value, label }) => (
              <SelectItem value={value} className="rounded-lg" key={value}>
                {label}
              </SelectItem>
            ))}
          </SelectContent>
        </Select>
      </CardHeader>
      <CardContent className="px-2 sm:p-6">
        <ChartContainer
          config={chartConfig}
          className="aspect-auto h-[500px] w-[1200px]"
        >
          <BarChart
            accessibilityLayer={true}
            data={sumPerDay}
            margin={{
              left: 12,
              right: 12,
            }}
          >
            <CartesianGrid vertical={false} />
            <XAxis
              dataKey="date"
              tickLine={false}
              axisLine={false}
              tickMargin={8}
              minTickGap={32}
              tickFormatter={(value) =>
                new Date(value).toLocaleDateString("en-US", {
                  month: "short",
                  day: "numeric",
                })
              }
            />
            <YAxis
              dataKey="kwh"
              tickLine={false}
              axisLine={false}
              tickMargin={8}
              minTickGap={32}
            />
            <ChartTooltip
              content={
                <ChartTooltipContent
                  className="w-[150px]"
                  nameKey="kwh"
                  labelFormatter={(value) =>
                    new Date(value).toLocaleDateString("en-US", {
                      month: "short",
                      day: "numeric",
                      year: "numeric",
                    })
                  }
                />
              }
            />
            <Bar dataKey="kwh" fill={chartConfig.kwh.color} />
          </BarChart>
        </ChartContainer>
      </CardContent>
    </Card>
  );
}
