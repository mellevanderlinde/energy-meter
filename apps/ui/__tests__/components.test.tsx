import { test, expect, vi, describe, beforeAll } from "vitest";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@/components/theme-provider";
import Chart from "@/components/chart";
import Home from "@/app/page";

describe("Components", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      value: vi.fn().mockImplementation(() => ({
        addListener: vi.fn(),
      })),
    });

    const ResizeObserverMock = vi.fn(() => ({
      observe: vi.fn(),
    }));

    vi.stubGlobal("ResizeObserver", ResizeObserverMock);
    vi.stubEnv("MOCK_DATA", "true");
  });

  test("Match theme provider", () => {
    const { container } = render(<ThemeProvider />);
    expect(container).toMatchSnapshot();
  });

  test("Match chart", () => {
    const { container } = render(<Chart />);
    expect(container).toMatchSnapshot();
  });

  test("Match home", () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
});
