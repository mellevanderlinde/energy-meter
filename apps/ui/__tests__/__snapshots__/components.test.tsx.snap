// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components > Match chart 1`] = `
<div>
  <div
    class="rounded-xl border bg-card text-card-foreground shadow"
  >
    <div
      class="flex-col p-6 flex items-center gap-2 space-y-0 border-b py-5 sm:flex-row"
    >
      <div
        class="grid flex-1 gap-1 text-center sm:text-left"
      >
        <div
          class="font-semibold leading-none tracking-tight"
        >
          Energy Usage
        </div>
        <div
          class="text-sm text-muted-foreground"
        >
          Showing kWh consumption over time
        </div>
      </div>
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-label="Select a value"
        class="flex h-9 items-center justify-between whitespace-nowrap border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[160px] rounded-lg sm:ml-auto"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Last 14 days
        </span>
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down h-4 w-4 opacity-50"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-6 pt-0 px-2 sm:p-6"
    >
      <div
        class="flex justify-center text-xs [&_.recharts-cartesian-axis-tick_text]:fill-muted-foreground [&_.recharts-cartesian-grid_line[stroke='#ccc']]:stroke-border/50 [&_.recharts-curve.recharts-tooltip-cursor]:stroke-border [&_.recharts-dot[stroke='#fff']]:stroke-transparent [&_.recharts-layer]:outline-none [&_.recharts-polar-grid_[stroke='#ccc']]:stroke-border [&_.recharts-radial-bar-background-sector]:fill-muted [&_.recharts-rectangle.recharts-tooltip-cursor]:fill-muted [&_.recharts-reference-line_[stroke='#ccc']]:stroke-border [&_.recharts-sector[stroke='#fff']]:stroke-transparent [&_.recharts-sector]:outline-none [&_.recharts-surface]:outline-none aspect-auto h-[500px] w-[1200px]"
        data-chart="chart-r1"
      >
        <style>
          
 [data-chart=chart-r1] {
  --color-kwh: hsl(var(--chart-1));
}


.dark [data-chart=chart-r1] {
  --color-kwh: hsl(var(--chart-1));
}

        </style>
        <div
          class="recharts-responsive-container"
          style="width: 100%; height: 100%; min-width: 0;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Components > Match home 1`] = `
<div>
  <div
    class="grid grid-rows-[20px_1fr_20px] items-center justify-items-center min-h-screen p-8 pb-20 gap-16 sm:p-20 font-[family-name:var(--font-geist-sans)]"
  >
    <main
      class="flex flex-col gap-8 row-start-2 items-center sm:items-start"
    >
      <div
        class="rounded-xl border bg-card text-card-foreground shadow"
      >
        <div
          class="flex-col p-6 flex items-center gap-2 space-y-0 border-b py-5 sm:flex-row"
        >
          <div
            class="grid flex-1 gap-1 text-center sm:text-left"
          >
            <div
              class="font-semibold leading-none tracking-tight"
            >
              Energy Usage
            </div>
            <div
              class="text-sm text-muted-foreground"
            >
              Showing kWh consumption over time
            </div>
          </div>
          <button
            aria-autocomplete="none"
            aria-controls="radix-:r5:"
            aria-expanded="false"
            aria-label="Select a value"
            class="flex h-9 items-center justify-between whitespace-nowrap border border-input bg-transparent px-3 py-2 text-sm shadow-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[160px] rounded-lg sm:ml-auto"
            data-state="closed"
            dir="ltr"
            role="combobox"
            type="button"
          >
            <span
              style="pointer-events: none;"
            >
              Last 14 days
            </span>
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-down h-4 w-4 opacity-50"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m6 9 6 6 6-6"
              />
            </svg>
          </button>
        </div>
        <div
          class="p-6 pt-0 px-2 sm:p-6"
        >
          <div
            class="flex justify-center text-xs [&_.recharts-cartesian-axis-tick_text]:fill-muted-foreground [&_.recharts-cartesian-grid_line[stroke='#ccc']]:stroke-border/50 [&_.recharts-curve.recharts-tooltip-cursor]:stroke-border [&_.recharts-dot[stroke='#fff']]:stroke-transparent [&_.recharts-layer]:outline-none [&_.recharts-polar-grid_[stroke='#ccc']]:stroke-border [&_.recharts-radial-bar-background-sector]:fill-muted [&_.recharts-rectangle.recharts-tooltip-cursor]:fill-muted [&_.recharts-reference-line_[stroke='#ccc']]:stroke-border [&_.recharts-sector[stroke='#fff']]:stroke-transparent [&_.recharts-sector]:outline-none [&_.recharts-surface]:outline-none aspect-auto h-[500px] w-[1200px]"
            data-chart="chart-r6"
          >
            <style>
              
 [data-chart=chart-r6] {
  --color-kwh: hsl(var(--chart-1));
}


.dark [data-chart=chart-r6] {
  --color-kwh: hsl(var(--chart-1));
}

            </style>
            <div
              class="recharts-responsive-container"
              style="width: 100%; height: 100%; min-width: 0;"
            />
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`Components > Match theme provider 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;{let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&m?d.map(f=&gt;m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("data-theme","theme","system",null,["light","dark"],null,true,true)
  </script>
</div>
`;
